ARCHITECTURE.md

```diff
@@ -0,0 +1,275 @@
+Bob  : 0  <-------------------/
+```
+
+Because Alice's latest feed entry refers to Bob's latest feed entry, there is
+now only one "head" in the database. That means there is enough information in
+Alice's seq=3 entry to find any other key in the database. In the last example,
+there were two heads (Alice's seq=2 and Bob's seq=0); both of which would need
+to be read internally in order to locate any key in the database.
+
+## Authorization
+
+The set of hypercores are *authorized* in that the original author of the first
+hypercore in a hyperdb must explicitly denote in their append-only log that the
+public key of a new hypercore is permitted to edit the database. Any authorized
+member may authorize more members. There is no revocation or other author
+management elements currently.
+
+## Incremental index
+
+HyperDB builds an *incremental index* with every new key/value pairs ("nodes")
+written. This means a separate data structure doesn't need to be maintained
+elsewhere for fast writes and lookups: each node written has enough information
+to look up any other key quickly and otherwise navigate the database.
+
+Each node stores the following basic information:
+
+- `key`: the key that is being created or modified. e.g. `/home/sww/dev.md`
+- `value`: the value stored at that key.
+- `seq`: the sequence number of this entry in the owner's hypercore. 0 is the
+  first, 1 the second, and so forth.
+- `feed`: the ID of the hypercore writer that wrote this
+- `path`: a 2-bit hash sequence of the key's components
+- `trie`: a navigation structure used with `path` to find a desired key
+- `clock`: vector clock to determine node insertion causality
+- `feeds`: an array of { feedKey, seq } for decoding a `clock`
+
+### Vector clock
+
+Each node stores a [vector clock](https://en.wikipedia.org/wiki/Vector_clock) of
+the last known sequence number from each feed it knows about. This is what forms
+the DAG structure.
+
+For example, Alice's vector clock on her seq=3 entry above would be `[3, 0]`
+since she knows of her latest entry (seq=3) and Bob's (seq=0).
+
+*TODO: is this correct? ^^^*
+
+### Prefix trie
+
+Given a HyperDB with hundreds of entries, how can a key like `/a/b/c` be looked
+up quickly?
+
+Each node stores a *prefix [trie](https://en.wikipedia.org/wiki/Trie)* that
+assists with finding the shortest path to the desired key.
+
+When a node is written, its *prefix hash* is computed. This done by first
+splitting the key into its components (`a`, `b`, and `c` for `/a/b/c`), and then
+hashing each component into a 32-character hash, where one character is a 2-bit
+value (0, 1, 2, or 3). The `prefix` hash for `/a/b/c` is
+
+```js
+node.path = [
+1, 2, 0, 1, 2, 0, 2, 2, 3, 0, 1, 2, 1, 3, 0, 3, 0, 0, 2, 1, 0, 2, 0, 0, 2, 0, 0, 3, 2, 1, 1, 2,
+0, 1, 2, 3, 2, 2, 2, 0, 3, 1, 1, 3, 0, 3, 1, 3, 0, 1, 0, 1, 3, 2, 0, 2, 2, 3, 2, 2, 3, 3, 2, 3,
+0, 1, 1, 0, 1, 2, 3, 2, 2, 2, 0, 0, 3, 1, 2, 1, 3, 3, 3, 3, 3, 3, 0, 3, 3, 2, 3, 2, 3, 0, 1, 0,
+4 ]
+```
```

Comment:
(on smaller screens there are more newlines than intended because of wrapping so it might be helpful to denote newline explicitly like with `\n` for example. i was reading this on my phone and was confused at this point.)